### PR TITLE
Fix onScrollEnd being called with outdated values (Fixes #226)

### PR DESCRIPTION
**Pull request by pradeepnschrodinger**
## Description

onScrollEnd made use of this.state to get values like firstRowIndex
React doesn't guarantee that this.state contains the updated values, but it does provide an API where a callback will be called after state is actually handled.

Fixed by making use of setState callback.
Also fixed scrollStop being called for every render as a side effect of this.

## Motivation and Context

Fixes the problem mentioned in #355
SS-26704: Scrolling with scroll bar in large LR leaves cells pending forever

## How Has This Been Tested?

Tested this in our examples. Made sure to mimic LD by using controlled scrolling through onVerticalScroll hack.
@wcjordan , Tim Choi, we'll probably need to test this on LD and see if SS-26704 gets fixed.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: src/FixedDataTable.js [hunk 1]
```diff
@@ -535,11 +535,10 @@ var FixedDataTable = createReactClass({
       this._didScrollStart();
     }
 
-    // Cancel any pending debounced scroll handling and handle immediately.
-    this._didScrollStop.reset();
-    this._didScrollStopSync();
-
-    this.setState(this._calculateState(nextProps, this.state));
+    this.setState(this._calculateState(nextProps, this.state), () => {
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: so now `didScrollStop` is only called after `setState` actually updates `this.state`
I also removed the reset call here (see #398)

File: src/FixedDataTable.js [hunk 2]
```diff
@@ -1321,13 +1323,13 @@ var FixedDataTable = createReactClass({
...
-        this.setState({
+        newState = {
           scrollX: x,
-        });
+        };
       }
     }
 
-    this._didScrollStop();
+    this.setState(newState, this._didScrollStop);
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: same logic as explained above: call `didScrollStop` only after `this.state` is updated

File: src/FixedDataTable.js [hunk 3]
```diff
@@ -1338,13 +1340,16 @@ var FixedDataTable = createReactClass({
...
+
     var onHorizontalScroll = this.props.onHorizontalScroll;
     if (onHorizontalScroll ? onHorizontalScroll(scrollPos) : true) {
-      this.setState({
-        scrollX: scrollPos,
-      });
+      newState.scrollX = scrollPos;
+      this.setState(newState, this._didScrollStop);
+    } else {
+      this._didScrollStop();
```
Comment (pradeepnschrodinger) [inline src/FixedDataTable.js]: @wcjordan , we don't update `scrollX` if `onHorizontalScroll` returns false. This means we can directly call `didScrollStop`

File: src/FixedDataTable.js (hunk 1 above)
Comment (wcjordan) [reply src/FixedDataTable.js]: nice!  

The comment is out of date now though, right?  We're not cancelling any pending scroll stop, we're just scheduling one.

**Comment by wcjordan** [comment]
It looks like this targets master which is v0.8.22, rather than v0.8.3... how are you planning to navigate that?  Is this PR just for review and not for merging?

**Comment by pradeepnschrodinger** [comment]
> this, lgtm. Are you planning to add the feature flag before merging? I'll wait til then to fully approve.

Great, sounds good. I'll fix the comment and put the FF as a prop now (maybe a prop named `fixOnScrollEnd`).



> It looks like this targets master which is v0.8.22, rather than v0.8.3... how are you planning to navigate that? Is this PR just for review and not for merging?

My bad, should have mentioned here. This actually targets v0.8.3 (see [compare](https://github.com/schrodinger/fixed-data-table-2/compare/v0.8.3...master-0.8.3-Fix-OnScrollEnd)). I couldn't figure out how to make a PR with base as a specific tag..

We'll merge it right you approve it.

**Comment by wcjordan** [comment]
I think you can create a fresh branch (say 0.8.3-LD) from the 0.8.3 tag and push it.  Then do a PR off that branch

**Comment by pradeepnschrodinger** [comment]
> I think you can create a fresh branch (say 0.8.3-LD) from the 0.8.3 tag and push it. Then do a PR off that branch

Done!

**Pull request merged**